avm/res/network/virtual-network/main.bicep

```diff
@@ -423,3 +414,133 @@ type subnetType = {
+  | 11 // 2,097,152 addresses
+  | 12 // 1,048,576 addresses
+  | 13 // 524,288 addresses
+  | 14 // 262,144 addresses
+  | 15 // 131,072 addresses
+  | 16 // 65,536 addresses
+  | 17 // 32,768 addresses
+  | 18 // 16,384 addresses
+  | 19 // 8,192 addresses
+  | 20 // 4,096 addresses
+  | 21 // 2,048 addresses
+  | 22 // 1,024 addresses
+  | 23 // 512 addresses
+  | 24 // 256 addresses
+  | 25 // 128 addresses
+  | 26 // 64 addresses
+  | 27 // 32 addresses
+  | 28 // 16 addresses
+  | 29 // 8 addresses
+  | 30 // 4 addresses
+  | 31 // 2 addresses
+
+@sealed()
+@export()
+type ipamPoolPrefixAllocationsType = {
+  @description('Required. The resource ID of the IPAM pool to allocate the address prefix from.')
+  ipamPoolResourceId: string
+
+  @description('Required. The CIDR prefix size to allocate from the IPAM pool (e.g., `24` for a /24 subnet with 256 addresses).')
+  cidr: cidrPrefixType
+}
+
+@sealed()
+@export()
+type ipamAddressPrefixesType = {
+  @description('Required. The allocation method for the address prefix. Must be set to `ipam` for IPAM-based allocation.')
+  by: 'ipam'
+
+  @description('Required. Array of IPAM pool prefix allocations specifying which pools to allocate address space from.')
+  ipamPoolPrefixAllocations: ipamPoolPrefixAllocationsType[]
+}
+
+@sealed()
+@export()
+type addressPrefixesType = {
+  @description('Required. The allocation method for the address prefix. Must be set to `addressPrefixes` for manually specified address prefixes.')
+  by: 'addressPrefixes'
+
+  @description('Required. Array of manually specified address prefixes in CIDR notation (e.g., `["10.0.0.0/16", "10.1.0.0/16"]`).')
+  addressPrefixes: string[]
+}
+
+@sealed()
+@export()
+type addressPrefixType = {
+  @description('Required. The allocation method for the address prefix. Must be set to `addressPrefix` for a single manually specified address prefix.')
+  by: 'addressPrefix'
+
+  @description('Required. The manually specified address prefix in CIDR notation (e.g., `10.0.0.0/24`).')
+  addressPrefix: string
+}
+
+@export()
+@discriminator('by')
+@description('Discriminated union type for Virtual Network address space configuration. Supports either IPAM-based allocation or manually specified address prefixes.')
+type addressSpaceType = ipamAddressPrefixesType | addressPrefixesType
+
+@export()
+@discriminator('by')
+@description('Discriminated union type for Subnet address configuration. Supports IPAM-based allocation, multiple address prefixes, or a single address prefix.')
+type subnetAddressSpaceType = ipamAddressPrefixesType | addressPrefixesType | addressPrefixType
+
+@export()
+type keyValuePairType = {
+  @description('Required. A key-value pair. (e.g., Environment: Production)')
+  *: string
+}
+
+// Functions to get number of hosts in a CIDR prefix
+@description('Returns a lookup table mapping CIDR prefixes to their corresponding number of IP addresses.')
+func getCidrHostCounts() keyValuePairType => {
```

Comment:
Hey @duanany,
would you mind if we publish those functions also in a dedicated `utl` (utility) module to make them available beyond the scope of this resource? I'm not sure but I could imagine that users may also benefit from it in other scopes... Just a thought